snuba/utils/streams/consumers/backends/synchronized.py

```diff
@@ -0,0 +1,271 @@
+        return self.__backend.subscribe(
+            topics, on_assign=assignment_callback, on_revoke=on_revoke
+        )
+
+    def unsubscribe(self) -> None:
+        # Update the subscription, so that the commit log consumer can stop
+        # consuming the commit log data for the previous (and no longer
+        # potentially assigned) topics.
+        with self.__lock:
+            self.__subscription = Subscription([])
+
+        return self.__backend.unsubscribe()
+
+    def poll(
+        self, timeout: Optional[float] = None
+    ) -> Optional[Message[TStream, TOffset, TValue]]:
+        # Check to see if any streams are paused that are able to be resumed
+        # based on the progression of the remote offset. (This could go the
+        # other way as well -- remote offsets may be rolled back -- but those
+        # will be caught later during normal consumption. The only benefit to
+        # pausing them here would be to avoid buffering messages that will
+        # likely be discarded. On the other hand, by the time we get to consume
+        # these messages, the offsets may have been synchronized at that point.
+        # It's hard to predict which scenario is more likely -- in the interest
+        # of buffering and low latency, it probably makes sense to avoid
+        # pausing here?)
+        with self.__lock:
+            # TODO: This should be limited to only streams that are currently
+            # paused due to the remote offsets -- we should even be able to get
+            # those streams without acquiring the lock here.
+            for stream, offset in self.__backend.tell().items():
+                effective_remote_offset = self.__get_effective_remote_offset(stream)
+                if (
+                    effective_remote_offset is not None
+                    and effective_remote_offset > offset
+                ):
+                    # TODO: This should be more intelligent (see comment in the
+                    # assignment callback in ``subscribe`` for more details.)
+                    self.__backend.resume([stream])
+
+        message = self.__backend.poll(timeout)
+        if message is None:
+            return None
+
+        # Check to ensure that consuming this message will not cause the local
+        # offset to exeed the remote offset(s).
+        with self.__lock:
+            effective_remote_offset = self.__get_effective_remote_offset(message.stream)
+            if (
+                effective_remote_offset is None
+                or message.offset >= effective_remote_offset
+            ):
+                # If this message does exceed the remote offset(s), it should be
+                # dropped, the stream should be paused, and the offset should be
+                # reset so that this message is consumed again when the stream is
+                # resumed.
+                self.__backend.pause([message.stream])
+                self.__backend.seek({message.stream: message.offset})
```

Comment:
This chunk is really important: this is the only thing keeping the consumer from reading into parts of the topic it shouldn't be.

This actually handles the failure scenario a bit more elegantly than the Sentry one does if offsets jump ahead unexpectedly: [that one just fails offset validation and crashes](https://github.com/getsentry/sentry/blob/2f9169c41d7106d1d438502bda017eb965364a56/src/sentry/eventstream/kafka/consumer.py#L335-L341).